src/wpool/mongoose_gun_worker.erl

```diff
@@ -0,0 +1,262 @@
+    end;
+%% `gun_down' is a message informing that Gun has lost the connection.
+%% It may try to reconnect, according to the `retry' and `retry_timeout' options,
+%% passed in the init options. They default to 5 retries, each with 5 second
+%% timeout.
+handle_info({gun_down, PID, Protocol, Reason, KilledStreams, UnprocessedStreams}, State) ->
+    ?WARNING_MSG("gun_down in mongoose_gun_worker. "
+                 "Gun has lost the ~p connection ~p because of \"~p\", killing ~p streams.",
+                 [Protocol, PID, Reason, length(KilledStreams) + length(UnprocessedStreams)]),
+    ?DEBUG("gun_down in mongoose_gun_worker. Killed streams: ~p. Unprocessed streams: ~p",
+           [KilledStreams, UnprocessedStreams]),
+    {noreply, State};
+%% `gun_up' is a message informing that Gun has reconnected after a `gun_down'.
+handle_info({gun_up, ConnPid, _Protocol},
+            State = #state{pid = ConnPid}) ->
+    ?DEBUG("gun_up in mongoose_gun_worker. Connection is back up with PID: ~p", [ConnPid]),
+    {noreply, State};
+%% `gun_error' is a message informing about any errors concerning connections or
+%% streams handled by Gun.
+handle_info({gun_error, ConnPid, Reason},
+            State = #state{pid = ConnPid}) ->
+    ?WARNING_MSG("gun_error in mongoose_gun_worker. Reason: ~p.", [Reason]),
+    {noreply, State};
+handle_info({gun_error, ConnPid, StreamRef, Reason},
+            State = #state{pid = ConnPid}) ->
+    ?WARNING_MSG("gun_error in mongoose_gun_worker. Reason: ~p. Stream reference: ~p",
+                 [Reason, StreamRef]),
+    {noreply, State};
+%% After `retry' number of `gun_down' messages, the connection process dies.
+%% Because it is monitored, we receive a `DOWN' message and may restart the
+%% connection.
+handle_info({'DOWN', MRef, process, ConnPid, Reason},
+            #state{pid = ConnPid, monitor = MRef} = State) ->
+    ?WARNING_MSG("Mongoose_gun_worker has lost the connection with PID ~p. Reason ~p.",
+                 [ConnPid, Reason]),
+    ConnectedState = open_connection(State),
+    NewState = retry_all(ConnectedState),
+    {noreply, NewState};
+handle_info(M, S) ->
+    ?ERROR_MSG("Unexpected message in gun_worker ~p", [M]),
+    {noreply, S}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+parse_uri(Host) ->
+    M = uri_string:parse(Host),
+    H = maps:get(host, M, "") ++ maps:get(path, M, ""),
+    Tls = case maps:get(scheme, M, undefined) of
+        "https" -> #{transport => tls};
+              _ -> #{}
+    end,
+    {H, maps:get(port, M, undefined), maps:merge(default_opts(), Tls)}.
+
+open_connection(State) ->
+    {ok, PID} = gun:open(State#state.host, State#state.port, State#state.opts),
+    {ok, Protocol} = gun:await_up(PID),
+    State#state{pid = PID, protocol = Protocol, monitor = monitor(process, PID)}.
```

Comment:
```suggestion
    MRef = monitor(process, PID),
    {ok, Protocol} = gun:await_up(PID, MRef),
    State#state{pid = PID, protocol = Protocol, monitor = MRef}.
```
Notice that if this is not done, `gun:await_up/1` itself is exactly doing a monitor-await-demonitor, and then we monitor again. Just saving some redundancy.